### PR TITLE
Apply a floor to the calculated next lease manager tick

### PR DESCRIPTION
**Pull request by manadart**
If multiple entities fail to renew their lease, and their expiries are less than a second apart, a tick in the lease manager can cause leases to be checked, but no work to be done.

This is because expired leases are only removed from the Raft FSM upon ticks of the global clock updater I.e. at 1 second intervals.

Reading the leases acquires a read lock in the FSM, so we want to minimise accessing them if there might be nothing to do.

Here we simple add a second to any calculated `nextTick` time, ensuring that we reassess blocks no sooner than a second from the last check.

## QA steps

- Bootstrap to LXD.
- `juju model-config -m controller logging-config="<root>=INFO;juju.worker.lease.raft=TRACE;juju.core.raftlease=TRACE;juju.worker.globalclockupdater.raft=TRACE"`.
- `for id in {1..20}; do juju deploy ubuntu "u$id"; done` and wait for quiescence.
- Stop the containers with a 0.5 second pause between (use your model ID for container names): `for id in {0..19}; do lxc stop "juju-034fd1-$id"; sleep 0.5; done`.
- Wait a couple of minutes and run `juju debug-log -m controller --replay | grep 'next expire'`.
- There should be no intervals less than 1s. Lowest should look like this:
```
machine-0: 15:10:42 TRACE juju.worker.lease.raft [dd150c] next expire in 1s 2021-08-12 13:10:43.42272571 +0000 UTC m=+2173.880530472
```

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Lease manager - apply floor to block check interval](https://trello.com/c/auDn6lwL/3870-lease-manager-apply-floor-to-block-check-interval)

File: worker/lease/manager.go [hunk 1]
```diff
@@ -520,27 +519,30 @@ func (manager *Manager) checkBlocks(blocks blocks) {
...
 		nextTick = info.Expiry
 	}
-	manager.config.Logger.Tracef("[%s] next expire in %v %v",
-		manager.logContext, nextTick.Sub(now).Round(time.Millisecond), nextTick)
+
+	// The lease clock is ticked every second, so sub-second durations
+	// can cause block checks resulting in no action. This is wasteful because
+	// accessing leases from the FSM causes read-locks, potentially
+	// contributing to contention. Round *up* to the nearest second.
+	nextDuration := (nextTick.Sub(now) + 500*time.Millisecond).Round(time.Second)
```
Comment (jameinel) [inline worker/lease/manager.go:543]: I wonder about finding the min, and just always adding 1s to give time for things to batch a bit more. It won't really be problematic and might allow a little bit more batching.

Given that the only thing that actually expires the leases is coordination with the FSM there will always be a bit of estimation error. eg we think the lease will expire in 950ms but the clock won't actually tick for at least 1s, and if we wake up right at 1s, but it ticks (plus FSM sync time) it doesn't wake up until 1.1s and then we miss, and end up hitting the "next expire before now" case. And if you hit that acase 

I wonder if we would actually want:

max(now, min(all_expiries_even_in_the_past)) + 1s

Thoughts?

Comment (manadart) [reply worker/lease/manager.go]: This is done.

**Review by hmlanigan — APPROVED** [review]
QA went well, a few at 967ms, but that's close enough per discussion.

**Comment by manadart** [comment]
$$merge$$

File: worker/lease/manager.go [hunk 2]
```diff
@@ -515,38 +513,38 @@ func (manager *Manager) checkBlocks(blocks blocks) {
...
+	}
+
+	// The lease clock ticks *at least* a second from now. Expirations only
+	// occur when the global clock updater ticks the clock, so this avoids
+	// too frequently checking with the potential of having no work to do.
+	// The blanket addition of a second is no big deal.
+	nextTick.Add(time.Second)
+
+	nextDuration := nextTick.Sub(now).Round(time.Millisecond)
+	manager.config.Logger.Tracef("[%s] next expire in %v %v", manager.logContext, nextDuration, nextTick)
```
Comment (jameinel) [inline worker/lease/manager.go:546]: The only thing I can think of for how we are seeing 
```
2021-08-18 19:59:52 TRACE juju.worker.lease.raft manager.go:546 [0a61b5] next expire in 966ms 2021-08-18 19:59:53.397286643 +0000 UTC m=+15976.527887156
```

the only thing I could think of is if there is a discrepancy between monotonic clock and global clock of a given time.

**Pull request merged**